### ISSUE TITLE
Update parameter to be closest to object type

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#473]
Think about you have a function parameter that expects a `Message`, and on each message we have a `User` (ex. `$message->user`). Now think that this function use *only* the `User` on `Message`, and never use another property or method from `Message`. So make more sense we update the parameter to require an `User`, and not a `Message`. Making this change allow our method be more flexible, once that you use more weaker type (_let's just say it_).

**`Message` class structure example:**

```php
class Message {
    /** @var User */   public $user;
    /** @var Carbon */ public $date;
}
```

**Case 1:** you are using more than one property/method from `Message`. In this case, the inspection should not be activated.

```php
function example(Message $message) {
    return $message->date->format('Y-m-d') . ' - ' . $message->user->name();
}

example($message);
```

**Case 2:** you are using only the `User` on `Message`, how many times you need.

```php
// Warning: parameter $message type could be modified from Message to User.
function example(Message $message) {
    return $message->user->name() . ' - ' . $message->user->age();
}

example($message);
```

**Fix**:

```php
function example(User $user) {
    return $user->name() . ' - ' . $user->age();
}

example($message->user);
```

**Comment by rentalhost** [comment]
I think that it is a good feature to be implemented, but because PHP doesn't supports method overloading maybe it could causes a lot of false positives in too early development phasis. For instance, the `example(Message $message)` could not be useful in the initial phase, but certainly will be useful on next phases. By otherside, I have a lot of method in "definitive" version that could be simplified.

**Comment by funivan** [comment]
@rentalhost you describe [Law_of_Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) =)
